llvm/lib/Target/X86/X86AsmPrinter.cpp

```diff
@@ -216,24 +216,77 @@ void processInstructions(
 }
 
 /// Walk through the control flow of a function starting at the entry block
-/// using depth first search. This makes sure the mappings are are generated in
-/// the correct order.
+/// using depth first search and analyse spills into other registers or the
+/// stack. This is required in order to encode additional locations into
+/// stackmaps. When blocks appear in a loop, we run a fixed-point algorithm to
+/// make sure that extra locations are valid in each iteration.
+///
+/// To give an example:
+///
+/// bb0:
+///   mov rax, [rbp-8]
+///   br bb1
+///
+/// bb1:
+///   STACKMAP [rax]
+///   br bb2
+///
+/// bb2:
+///   mov rax, [rbp-16]
+///   br bb1
+///
+/// Here, the first time we enter `bb1` the variable that's stored in `rax` has
+/// an extra location `rbp-8`. However, when we enter `bb1` again via `bb2`,
+/// the extra location has now changed to `rbp-16`. We thus encode neither
+/// extra location into the stackmap.
+///
 /// YKFIXME: Can be updated to use an iterative approach.
 void findSpillLocations(
     const MachineBasicBlock *MBB,
-    std::set<const MachineBasicBlock *> &Seen,
+    std::map<const MachineBasicBlock *, std::map<Register, std::set<int64_t>>> &MBSpillMaps,
     std::map<Register, std::set<int64_t>> SpillMap,
     std::map<const MachineInstr *, std::map<Register, std::set<int64_t>>> &StackmapSpillMaps
     ) {
 
-  // Block has already been processed.
-  if (Seen.count(MBB) > 0) {
-    return;
+  int changed = false;
+  // For each livein intersect the existing spills for this block with the
+  // spills passed down from the predecessor block.
```

Comment:
Just checking, this works for blocks with more than a single predecessor, right? If so, pluralise `block` above.

Reply:
Yes. Though this processed one predecessor at a time. So in this moment there's only one spillmap passed down from one predecessor.